### PR TITLE
Use a more compatible serialization of bytes

### PR DESCRIPTION
**Pull request by aidanhs**
Based on the Python 3 serialization of byte strings in proto v2

### CODE REVIEW

File: src/ser.rs [hunk 1]
```diff
@@ -462,16 +462,41 @@ impl<'a, W: io::Write> ser::Serializer for &'a mut Serializer<W> {
...
+            // string match their codepoints. So converting to utf8 encoding is
+            // as simple as interpreting each byte as a unicode codepoint and
+            // then encoding as utf8 - https://stackoverflow.com/a/28175593/2352259
+            let utf8_value: String = value.iter().map(|&c| c as char).collect();
+            self.writer.write_u32::<LittleEndian>(utf8_value.len() as u32)?;
+            self.writer.write_all(utf8_value.as_bytes())?;
+            const LATIN1: &[u8] = b"latin1";
+            self.write_opcode(BINUNICODE)?;
+            self.writer.write_u32::<LittleEndian>(LATIN1.len() as u32)?;
+            self.writer.write_all(LATIN1)?;
```
Comment (birkenfeld) [inline src/ser.rs]: Could this use two calls of self.serialize_str() instead of inlining it?

**Review by birkenfeld — COMMENTED** [review]
Thanks, except for a small simplification this looks good to me.

I think this will need a major version bump for the behavior change, and since nothing else has come up for a long time, I'm tempted to call it 1.0.

Are you aware of anything else you'd want changed for 1.0?

File: src/ser.rs (hunk 1 above)
Comment (aidanhs) [reply src/ser.rs]: Yep, great point.

**Comment by aidanhs** [comment]
I think the main thing I'd probably want to get in is the optimisation mentioned in the TODO, to use the memo for the two fixed strings - I'd probably put that in a major bump for safety because it influences how friendly the serializer is to other things serializing to the same pickle.

I'll probably get to in in the next couple of weeks if that works?

**Comment by birkenfeld** [comment]
Sounds good, I'll do the 1.0 then!

**Review by birkenfeld — APPROVED** [review]

**Comment by aidanhs** [comment]
Just wanted to follow up on this - unfortunately I'm not using serde-pickle any more and so won't get round to adding that optimisation. Sorry!

**Pull request merged**